File: packages/dart/lib/Seguranca/Ataques/AtaqueAdulteracao.dart
import 'dart:typed_data';

import 'package:criptografia/Seguranca/AlvoCriptografico.dart';
import 'package:criptografia/Seguranca/Ataque.dart';
import 'package:criptografia/Seguranca/CriptografiaAlvo.dart';
import 'package:criptografia/Seguranca/ResultadoAtaque.dart';
import 'package:criptografia/Seguranca/SkipAtaqueException.dart';
import 'package:criptografia/Seguranca/Util.dart';

/// O ataque mais importante da suíte: se um bit flipado em QUALQUER campo do
/// token (ciphertext, iv, ou o campo de integridade) ainda decifra "com
/// sucesso", a cifra não está autenticando o conteúdo.
class AtaqueAdulteracao extends Ataque {
  final int _tentativas;

  AtaqueAdulteracao({this._tentativas = 500});

  @override
  String nome() => 'Adulteração de bits (integridade)';

  @override
  ResultadoAtaque executar(AlvoCriptografico alvo) {
    if (alvo is! CriptografiaAlvo) {
      throw SkipAtaqueException('Precisa de base64url_encode/decode do alvo.');
    }

    final aceitosIndevidamente = <String>[];

    for (var t = 0; t < _tentativas; t++) {
      final texto = 'MSG_$t${repetir('X', randomInt(0, 50))}';
      final token = alvo.encrypt(texto);
      final decodificado = alvo.base64urlDecode(
        token.substring(alvo.prefixo().length),
      );
      final campos = alvo.decompor(decodificado);

      final nomeCampo = arrayRandKey(campos);
      final valor = Uint8List.fromList(campos[nomeCampo]!);
      if (valor.isEmpty) {
        continue;
      }
      final pos = randomInt(0, valor.length - 1);
      valor[pos] = valor[pos] ^ (1 << randomInt(0, 7));
      campos[nomeCampo] = valor;

      final tokenAdulterado =
          alvo.prefixo() + alvo.base64urlEncode(alvo.recompor(campos));

      try {
        final resultado = alvo.decrypt(tokenAdulterado);
        aceitosIndevidamente.add(
          'campo=$nomeCampo, texto original=$texto, resultado aceito=$resultado',
        );
      } catch (_) {
        // esperado - a adulteração deveria ser rejeitada
      }
    }

    if (aceitosIndevidamente.isNotEmpty) {
      return ResultadoAtaque(
        nome(),
        true,
        Severidade.critica,
        '${aceitosIndevidamente.length} de $_tentativas tokens adulterados '
            'foram ACEITOS',
        {'exemplos': aceitosIndevidamente.take(5).toList()},
      );
    }

    return ResultadoAtaque(
      nome(),
      false,
      Severidade.info,
      'Todas as $_tentativas adulterações foram rejeitadas',
    );
  }
}

File: packages/julia/src/Seguranca/Ataques/AtaqueAdulteracao.jl
"""
O ataque mais importante da suíte: se um bit flipado em QUALQUER campo do
token (ciphertext, iv, ou o campo de integridade) ainda decifra "com sucesso",
a cifra não está autenticando o conteúdo.
"""
struct AtaqueAdulteracao <: Ataque
    tentativas::Int
end

AtaqueAdulteracao(; tentativas::Integer = 500) = AtaqueAdulteracao(Int(tentativas))

nome(::AtaqueAdulteracao) = "Adulteração de bits (integridade)"

function executar(a::AtaqueAdulteracao, alvo::AlvoCriptografico)
    if !(alvo isa CriptografiaAlvo)
        throw(SkipAtaqueException("Precisa de base64url_encode/decode do alvo."))
    end

    aceitos = String[]

    for t in 0:(a.tentativas-1)
        texto = "MSG_$t" * repeat("X", random_int(0, 50))
        token = encrypt(alvo, texto)
        decodificado = base64url_decode(alvo, token[4:end])
        campos = decompor(alvo, decodificado)

        nome_campo = array_rand_key(campos)
        valor = copy(campos[nome_campo])
        isempty(valor) && continue
        pos = random_int(1, length(valor))
        valor[pos] = valor[pos] ⊻ UInt8(1 << random_int(0, 7))
        campos = merge(campos, NamedTuple{(nome_campo,)}((valor,)))

        token_adulterado = prefixo(alvo) * base64url_encode(alvo, recompor(alvo, campos))

        try
            resultado = decrypt(alvo, token_adulterado)
            push!(
                aceitos,
                "campo=$nome_campo, texto original=$texto, resultado aceito=$resultado",
            )
        catch
            # esperado — a adulteração deveria ser rejeitada
        end
    end

    if !isempty(aceitos)
        return ResultadoAtaque(
            nome(a),
            true,
            "critica",
            "$(length(aceitos)) de $(a.tentativas) tokens adulterados foram ACEITOS",
            Dict{String,Any}("exemplos" => aceitos[1:min(5, length(aceitos))]),
        )
    end

    return ResultadoAtaque(
        nome(a),
        false,
        "info",
        "Todas as $(a.tentativas) adulterações foram rejeitadas",
    )
end
